src/Util/ClassNameValue.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+/*
+ * This file is part of the Symfony MakerBundle package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\MakerBundle\Util;
+
+use Symfony\Bundle\MakerBundle\Str;
+
+/**
+ * @internal
+ */
+final class ClassNameValue
+{
+    private $fullClassName;
+    private $typeHint;
+
+    public function __construct(string $fullClassName, string $typeHint)
+    {
+        $this->fullClassName = $fullClassName;
+        $this->typeHint = $typeHint;
+    }
+
+    public function getShortName(): string
+    {
+        if ('self' === $this->typeHint) {
```

Comment:
The `self` case is (I believe) for cases when you're adding a relation that is from one class to *itself*. In that case, why would we not just use the short class name like normal? I think I'm missing something :).

Reply:
I was afraid to use `self::class`, then I'd it.

Reply:
Does `self::class()` work inside the relation annotations? If it does, I guess, yea, let's use it!

Reply:
I've tried to use `self` inside annotation but it doesn't work.

Reply:
Ok, no problem :).

So... is this `if` statement needed? I don't see how `self` is a special case. We pass the `$typeHint` as the second arg to the constructor. This comes from `ClassSourceManipulator`, where it has already done the logic to figure out the correct "short name" that's needed... including (I believe) when a relation is being made it itself. Can we remove this whole if block?

Reply:
Right, the `ClassSourceManipulator` resolve the `typeHint`, that can be `self` (when self relation `[1]`), class shortname (normal use `[2]`), or fqcn (when conflict `[3]`). Only in the case that the `typeHint` value is `self` we can't use it, we need to specify the class name (shortname for the self class).

```php
class Product {
    /** [1]
     * @var self
     * @ORM\ManyToOne(targetEntity=Product::class)
     */
    private $parent;

    /** [2]
     * @var Category
     * @ORM\ManyToOne(targetEntity=Category::class)
     */
    private $category;

    /** [3]
     * @var \App\Entity\Subdir\Category
     * @ORM\ManyToOne(targetEntity=\App\Entity\Subdir\Category::class)
     */
    private $subCategory;
}
```

BTW I can't figure out a better way to resolve it. :confused: